internal/service/domain/tree/tree.go

```diff
@@ -121,32 +124,43 @@ func (s *TreeService) Create(ctx context.Context, treeCreate *entities.TreeCreat
 		return nil, service.NewError(service.BadRequest, errors.Wrap(err, "validation error").Error())
 	}
 
-	fn := make([]entities.EntityFunc[entities.Tree], 0)
+	var treeClusterID *entities.TreeCluster
 	if treeCreate.TreeClusterID != nil {
-		treeClusterID, err := s.treeClusterRepo.GetByID(ctx, *treeCreate.TreeClusterID)
+		var err error
+		treeClusterID, err = s.treeClusterRepo.GetByID(ctx, *treeCreate.TreeClusterID)
 		if err != nil {
 			return nil, handleError(err)
 		}
-		fn = append(fn, tree.WithTreeCluster(treeClusterID))
 	}
 
+	var sensorID *entities.Sensor
 	if treeCreate.SensorID != nil {
-		sensorID, err := s.sensorRepo.GetByID(ctx, *treeCreate.SensorID)
+		var err error
+		sensorID, err = s.sensorRepo.GetByID(ctx, *treeCreate.SensorID)
 		if err != nil {
 			return nil, handleError(err)
 		}
-		fn = append(fn, tree.WithSensor(sensorID))
-	}
-
-	fn = append(fn,
-		tree.WithReadonly(treeCreate.Readonly),
-		tree.WithPlantingYear(treeCreate.PlantingYear),
-		tree.WithSpecies(treeCreate.Species),
-		tree.WithNumber(treeCreate.Number),
-		tree.WithLatitude(treeCreate.Latitude),
-		tree.WithLongitude(treeCreate.Longitude),
-	)
-	newTree, err := s.treeRepo.Create(ctx, fn...)
+	}
+
+	newTree, err := s.treeRepo.Create(ctx, func(tree *entities.Tree) (bool, error) {
+		tree.Readonly = treeCreate.Readonly
+		tree.PlantingYear = treeCreate.PlantingYear
+		tree.Species = treeCreate.Species
+		tree.Number = treeCreate.Number
+		tree.Latitude = treeCreate.Latitude
+		tree.Longitude = treeCreate.Longitude
+
+		// Apply TreeCluster and Sensor if available
+		if treeClusterID != nil {
+			tree.TreeCluster = treeClusterID
+		}
+		if sensorID != nil {
+			tree.Sensor = sensorID
+		}
```

Comment:
You can do all the pre-defining of the tree cluster and sensor within this function. Then you can assign them directly to the `tree` entity you get in this function and you can also remove the misleading names `treeClusterID` and `sensorID`